### PR TITLE
ConstProp: Correctly remove const if unknown value assigned to it.

### LINKED ISSUE #118328: `ConstProp` misoptimises pointer-typed enum field
Fuzzer generated MIR, reduced, and UB-free under Miri (for real this time :sweat_smile:)

```rust
#![feature(custom_mir, core_intrinsics)]
extern crate core;
use core::intrinsics::mir::*;

#[custom_mir(dialect = "runtime", phase = "initial")]
fn fn4() {
    mir! {
    let _1: isize;
    let _12: Adt55;
    let unit: ();
    {
    _12 = Adt55::Variant1 { fld0: 0, fld1: 0};
    SetDiscriminant(_12, 0);
    place!(Field::<*mut isize>(Variant(_12, 0), 0)) = core::ptr::addr_of_mut!(_1);
    Call(unit = fn19(Field::<*mut isize>(Variant(_12, 0), 0)), bb11, UnwindUnreachable())
    }
    bb11 = {
    Return()
    }
    }
}
#[custom_mir(dialect = "runtime", phase = "initial")]
pub fn fn19(mut _1: *mut isize) {
    mir! {
    {
    (*_1) = 1;
    Return()
    }
    }
}
pub fn main() {
    fn4();
    println!("here");
}
#[derive(Debug, Copy, Clone)]
pub enum Adt55 {
    Variant0 { fld0: *mut isize },
    Variant1 { fld0: u8, fld1: u64 },
}
```

Segfaults with `ConstProp` enabled:
```console
$ rustc -Zmir-opt-level=0 -Copt-level=0 -Zmir-enable-passes=+ConstProp repro.rs &&
 ./repro
Segmentation fault (core dumped)
```

`Field::<*mut isize>(Variant(_12, 0), 0))`, which is a valid pointer, somehow got propagated as 0:
```
// MIR for `fn4` before ConstProp

fn fn4() -> () {
    let mut _0: ();
    let mut _1: isize;
    let mut _2: Adt55;
    let mut _3: ();

    bb0: {
        _2 = Adt55::Variant1 { fld0: const 0_u8, fld1: const 0_u64 };
        discriminant(_2) = 0;
        ((_2 as variant#0).0: *mut isize) = &raw mut _1;
        _3 = fn19(((_2 as variant#0).0: *mut isize)) -> [return: bb1, unwind unreachable];
    }

    bb1: {
        return;
    }
}
```

```
// MIR for `fn4` after ConstProp

fn fn4() -> () {
    let mut _0: ();
    let mut _1: isize;
    let mut _2: Adt55;
    let mut _3: ();

    bb0: {
        _2 = Adt55::Variant1 { fld0: const 0_u8, fld1: const 0_u64 };
        discriminant(_2) = 0;
        ((_2 as variant#0).0: *mut isize) = &raw mut _1;
        _3 = fn19(const {0x0 as *mut isize}) -> [return: bb1, unwind unreachable];
    }

    bb1: {
        return;
    }
}
```

### PR DESCRIPTION
**Pull request by aDotInTheVoid**
Closes #118328

The problematic sequence of MIR is:

```rust
          _1 = const 0_usize;
          _1 = const _; // This is an associated constant we can't know before monomorphization.
          _0 = _1;
```

1. When `ConstProp::visit_assign` happens on `_1 = const 0_usize;`, it records that `0x0usize` is the value for `_1`.
2. Next `visit_assign` happens on `_1 = const _;`. Because the rvalue `.has_param()`, it can't be const evaled.
3. Finaly, `visit_assign` happens on `_0 = _1;`. Here it would think the value of `_1` was `0x0usize` from step 1.

The solution is to remove consts when checking the RValue fails, as they may have contained values that should now be invalidated, as that local was overwritten.

This should probably be back-ported to beta. Stable is more iffy, as it's gone unidentified since 1.70, so I only think it's worthwhile if there's another reason for a 1.74.1 release anyway.

### CODE REVIEW

File: tests/mir-opt/const_prop/issue_118328.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+// EMIT_MIR issue_118328.size_of.ConstProp.diff
+fn size_of<T>() -> usize {
+    let mut a = 0;
+    a = SizeOfConst::<T>::SIZE;
+    a
+}
+
+fn main() {
+    assert_eq!(size_of::<u32>(), std::mem::size_of::<u32>());
```
Comment (aDotInTheVoid) [inline tests/mir-opt/const_prop/issue_118328.rs:19]: I don't think this is actually running, and in general I'm not sure what the right way to write a robust test for this is.

**Comment by compiler-errors** [comment]
Please rebase out the second and fourth commits in this stack, since the latter undoes the former, and it's not particularly useful for git history that way. Either that, or just drop the last commit, since it's possibly useful to keep the instrumentation around, as long as it's debug or trace level.

**Comment by compiler-errors** [comment]
r=me when ci is green, since this seems like a pretty nasty regression

cc @cjgillot who may be curious about this const prop misoptimization, perhaps you have some ideas to make this code less fragile.

File: tests/mir-opt/const_prop/overwrite_with_const_with_params.rs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+// unit-test: ConstProp
+// compile-flags: -O
+// skip-filecheck
```
Comment (cjgillot) [inline tests/mir-opt/const_prop/overwrite_with_const_with_params.rs:3]: Could you add the filecheck annotations? See other files for example.

**Review by cjgillot — COMMENTED** [review]
R=me with test fortified.

**Comment by cjgillot** [comment]
> perhaps you have some ideas to make this code less fragile.

Yes. Removing this pass entirely, replaced by gvn.

File: tests/mir-opt/const_prop/overwrite_with_const_with_params.rs (hunk 2 above)
Comment (aDotInTheVoid) [reply tests/mir-opt/const_prop/overwrite_with_const_with_params.rs]: done

**Comment by apiraino** [comment]
Beta backport accepted as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-11-30/near/405137383)

@rustbot label +beta-accepted

**Pull request merged**